migration/migrator.go

```diff
@@ -0,0 +1,347 @@
+	tracersAPI    *tracers.API
+	traceCfg      *tracers.TraceConfig
+	migratedRef   *core.MigratedRef
+
+	stopCh chan struct{}
+}
+
+func NewStateMigrator(backend ethBackend, tracersAPI *tracers.API) *StateMigrator {
+	db := backend.ChainDb()
+
+	allocPreimage, err := zkPreimageWithAlloc(db)
+	if err != nil {
+		log.Crit("Failed to read genesis alloc", "err", err)
+	}
+	return &StateMigrator{
+		backend: backend,
+		db:      db,
+		zkdb: trie.NewDatabase(db, &trie.Config{
+			Preimages:   true,
+			Zktrie:      true,
+			KromaZKTrie: backend.BlockChain().TrieDB().IsKromaZK(),
+		}),
+		mptdb:         trie.NewDatabase(db, &trie.Config{Preimages: true}),
+		allocPreimage: allocPreimage,
+		tracersAPI:    tracersAPI,
+		traceCfg: &tracers.TraceConfig{
+			Tracer:       &tracerType,
+			TracerConfig: json.RawMessage(`{"diffMode": true}`),
+		},
+		migratedRef: core.NewMigratedRef(db),
+		stopCh:      make(chan struct{}),
+	}
+}
+
+func (m *StateMigrator) Start() {
+	log.Info("Start state migrator to migrate ZKT to MPT")
+	go func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		if m.migratedRef.Root() == (common.Hash{}) {
+			var safeHead *types.Header
+			if safe := m.backend.BlockChain().CurrentSafeBlock(); safe != nil {
+				safeHead = safe
+			} else {
+				safeHead = m.backend.BlockChain().Genesis().Header()
+			}
+			log.Info("Start migrate past state")
+			// Start migration from the head block. It takes long time.
+			err := m.migrateAccount(safeHead)
+			if err != nil {
+				log.Crit("Failed to migrate state", "error", err)
+			}
+
+			err = m.ValidateMigratedState(ctx, m.migratedRef.Root(), safeHead.Root)
+			if err != nil {
+				log.Crit("Migrated state is invalid", "error", err)
+			}
+			log.Info("Migrated past state have been validated")
+		}
+
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		log.Info("Start a loop to apply state of new block")
+		for {
+			select {
+			case <-ticker.C:
+				currentBlock := m.backend.BlockChain().CurrentSafeBlock()
+				// Skip block that have already been migrated.
+				if currentBlock == nil || m.migratedRef.BlockNumber() >= currentBlock.Number.Uint64() {
+					continue
+				}
+				if m.backend.BlockChain().Config().IsKromaMPT(currentBlock.Time) {
+					return
+				}
+				err := m.applyNewStateTransition(currentBlock.Number.Uint64())
+				if err != nil {
+					log.Error("Failed to apply new state transition", "error", err)
+				}
+			case <-m.stopCh:
+				cancel()
+				return
+			}
+		}
+	}()
+}
+
+func (m *StateMigrator) Stop() {
+	log.Info("Stopping state migrator")
+	close(m.stopCh)
+}
+
+func (m *StateMigrator) MigratedRef() *core.MigratedRef {
+	return m.migratedRef
+}
+
+func (m *StateMigrator) migrateAccount(header *types.Header) error {
+	log.Info("Migrate account", "root", header.Root, "number", header.Number)
+	startAt := time.Now()
+	var accounts atomic.Uint64
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
How about adding a global context for `StateMigrator` and use that here? + add `globalCtx.cancel()` in `Stop()` function.